### PR TITLE
Add New SelectiveMetadatAggregation method

### PR DESCRIPTION
**Pull request by eisenhauer**
Still a little tentative, want to make sure it doesn't break at high scales (on machines where gatherv breaks at high scales), but would like to merge at this point to get it into master.

### CODE REVIEW

**Comment by eisenhauer** [comment]
This does not yet change the default aggregation in BP5 file engine, it just adds an option.

**Review by anagainaru — COMMENTED** [review]
I don't see any evident problem with this. Is there any way to test the new aggregation schemes?

**Comment by eisenhauer** [comment]
This is still a bit of a work-in-progress.  I've got numbers on Frontier that shows the new scheme takes half the time as the old for kind-of-normal metadata (~100 vars, 100 arrays), and comparable time for E3SM-sized metadata.  But new involves full-scale gatherv, and Frontier seems to have unpredictable problems with this (sometimes slow, sometimes fast).  So not ready to finalize until that is sorted or at least understood.  Point taken about NewWriteMetadata.  Not sure about overload here though.  Will ponder.

**Comment by eisenhauer** [comment]
This needs to be rebased on top of the FFS upstream of PR #4391, then should pass MSAN.

**Comment by eisenhauer** [comment]
> This needs to be rebased on top of the FFS upstream of PR #4391, then should pass MSAN.

this has been done

**Comment by eisenhauer** [comment]
> Still a little tentative, want to make sure it doesn't break at high scales (on machines where gatherv breaks at high scales), but would like to merge at this point to get it into master.

IMHO, ready as default now.  Changed as appropriate.

File: source/adios2/engine/bp5/BP5Engine.h [hunk 1]
```diff
@@ -172,6 +172,8 @@ class BP5Engine
     MACRO(StatsBlockSize, SizeBytes, size_t, DefaultStatsBlockSize)                                \
     MACRO(Threads, UInt, unsigned int, 0)                                                          \
     MACRO(UseOneTimeAttributes, Bool, bool, true)                                                  \
+    MACRO(UseSelectiveMetadataAggregation, Bool, bool, true)                                       \
+    MACRO(OneLevelGatherSizeLimit, Int, int, 6000)                                                 \
```
Comment (pnorbert) [inline source/adios2/engine/bp5/BP5Engine.h:176]: `OneLevelGatherSizeLimit`: Size of what? To me, it first suggests a metadata size limit. Propose new names: `OneLevelGatherMPISizeLimit`, `OneLevelGatherNprocLimit`
Change could go into another PR with lines in the documentation ;-)

Comment (anagainaru) [reply source/adios2/engine/bp5/BP5Engine.h:176]: I would actually call it `OneLevelGatherRanksLimit` since the threshold is given by the number of ranks that are part of the gather right?

Comment (eisenhauer) [reply source/adios2/engine/bp5/BP5Engine.h]: I like OneLevelGatherRanksLimit.  I've changed to that, eliminated a couple of unused (and undocumented) parameters and added some docs on the parameters introduced here.  Turns out I miscounted and UseSelectiveMetadataAggregation was longer than anything and I had to reformat the table a bit to get it to fit.

**Review by anagainaru — APPROVED** [review]

**Review by pnorbert — APPROVED** [review]
